subprojects/base-services/src/main/java/org/gradle/internal/reflect/ReflectionCache.java

```diff
@@ -34,12 +34,14 @@ public T get(Class<?> key) {
         synchronized (lock) {
             cached = cache.get(key);
         }
-        if (cached != null) {
+        if (cached != null && !hasExpired(cached)) {
```

Comment:
Ok so I have found a way to make this fail. It involves a `System.gc()` right after the cached value has been validated, and it causes problems. I need to find a way to fix this without introducing strong references...